### PR TITLE
Refactoring to get ready for v3 state endpoint changes

### PR DESCRIPTION
**Pull request by toochevere**
These are the first set of changes (refactoring) for the State endpoint V3. 

Signed-off-by:  PJ <pj@balena.io>
Change-type: patch

### CODE REVIEW

File: src/features/device-state/routes/state.ts [hunk 1]
```diff
@@ -157,241 +313,185 @@ const stateQuery = _.once(() =>
...
+	filterDeviceConfig(config, device.os_version);
+	setMinPollInterval(config);
 
-		(device.manages__device as AnyObject[]).forEach((depDev) => {
-			const depAppId: number = depDev.belongs_to__application.__id;
-			const {
+	return config;
+};
+
+const getUserAppForState = async (device: any, config: Dictionary<string>) => {
```
Comment (Page-) [inline src/features/device-state/routes/state.ts:382]: Can you type this better than `any` please

File: src/features/device-state/utils.ts [hunk 2]
```diff
@@ -33,10 +33,14 @@ export const serviceInstallFromImage = (
 	}
 
 	let id: number;
-	if (typeof image.is_a_build_of__service === 'object') {
-		id = image.is_a_build_of__service.__id;
+	if (Array.isArray(image.is_a_build_of__service)) {
```
Comment (Page-) [inline src/features/device-state/utils.ts:36]: I think this is invalid without other changes to functionality in the code to `$expand` `is_a_build_of__service`? It also makes this a major change as this is part of the exported api surface of this module (and it gets used by cloud api in order to support the legacy state endpoint v1)

File: src/features/device-state/routes/state.ts [hunk 3]
```diff
@@ -25,6 +26,161 @@ export const varListInsert = (varList: EnvVarList, obj: Dictionary<string>) => {
...
+	name: string;
+	commit: string;
+	releaseId: number;
+	appId?: string;
+	uuid?: string;
+	services: {
+		[id: number]: AppService;
+	};
+	volumes: any;
+	networks: any;
```
Comment (Page-) [inline src/features/device-state/routes/state.ts:39]: Are these truly any or can they at least be tightened to eg AnyObject?

File: src/features/device-state/routes/state.ts [hunk 4]
```diff
@@ -25,6 +26,161 @@ export const varListInsert = (varList: EnvVarList, obj: Dictionary<string>) => {
...
+	apps: Dictionary<Partial<App>>;
+};
+
+async function buildAppFromRelease(
+	device: AnyObject,
+	application: AnyObject,
+	release: AnyObject,
+	config: Dictionary<string>,
+): Promise<App> {
+	let composition: any = {};
```
Comment (Page-) [inline src/features/device-state/routes/state.ts:64]: Why did you change this from `AnyObject | undefined` to `any`?

File: src/features/device-state/routes/state.ts [hunk 5]
```diff
@@ -157,241 +313,185 @@ const stateQuery = _.once(() =>
...
-					depAppState.releaseId = depRelease.id;
-					depAppState.imageId = image.id;
-					depAppState.commit = depRelease.commit;
-					depAppState.image = formatImageLocation(
-						image.is_stored_at__image_location,
-					);
-				}
-			},
-		);
+const getConfig = (device: any) => {
```
Comment (Page-) [inline src/features/device-state/routes/state.ts:367]: Can you type this better than `any` please

File: src/features/device-state/routes/state.ts [hunk 6]
```diff
@@ -157,241 +313,185 @@ const stateQuery = _.once(() =>
...
+				services: {},
+				networks: {},
+				volumes: {},
+				uuid: userAppFromApi.uuid,
+				appId: userAppFromApi.id,
+		  }
+		: await buildAppFromRelease(device, userAppFromApi, release, config);
+};
+
+const getDependent = async (device: any) => {
```
Comment (Page-) [inline src/features/device-state/routes/state.ts:401]: Can you type this better than `any` please

**Review by Page- — CHANGES_REQUESTED** [review]

File: src/features/device-state/utils.ts (hunk 2 above)
Comment (toochevere) [reply src/features/device-state/utils.ts:36]: revereted

File: src/features/device-state/routes/state.ts [hunk 7]
```diff
@@ -25,6 +26,161 @@ export const varListInsert = (varList: EnvVarList, obj: Dictionary<string>) => {
...
+	let composition: AnyObject = {};
+	const services: App['services'] = {};
+
+	// Parse the composition to forward values to the device
+	if (_.isObject(release.composition)) {
+		composition = release.composition;
+	} else {
+		try {
+			composition = JSON.parse(release.composition);
+		} catch (e) {
```
Comment (xginn8) [inline src/features/device-state/routes/state.ts:73]:
```suggestion
		} catch {
```

File: test/test-lib/fixtures.ts [hunk 8]
```diff
@@ -154,6 +199,104 @@ const loaders: Dictionary<LoaderFunc> = {
 			user,
 		});
 	},
+	release_tags: async (jsonData, fixtures) => {
```
Comment (xginn8) [inline test/test-lib/fixtures.ts:202]: can be dropped, duplicated above

File: src/features/device-state/routes/state.ts [hunk 9]
```diff
@@ -25,6 +26,161 @@ export const varListInsert = (varList: EnvVarList, obj: Dictionary<string>) => {
...
+			}
+
+			/**
+			 * Return a fake service install as this is what we will use for MultiApp
+			 */
+			return {
+				service: [image.is_a_build_of__service[0]],
+				device_service_environment_variable: [],
+			};
+		})();
```
Comment (Page-) [inline src/features/device-state/routes/state.ts:106]: This looks like functional changes rather than pure refactoring?

File: src/features/device-state/routes/state.ts [hunk 10]
```diff
@@ -25,6 +26,161 @@ export const varListInsert = (varList: EnvVarList, obj: Dictionary<string>) => {
...
+): Promise<App> {
+	let composition: AnyObject = {};
+	const services: App['services'] = {};
+
+	// Parse the composition to forward values to the device
+	if (_.isObject(release.composition)) {
+		composition = release.composition;
+	} else {
+		try {
+			composition = JSON.parse(release.composition);
```
Comment (Page-) [inline src/features/device-state/routes/state.ts:72]: fwiw I don't think this `JSON.parse` should ever be useful - I'd expect it to either already be parsed or potentially have had a null value in the db (although I don't think it's still non-nullable?). I'd leave it for a separate PR though along with verification that it's definitely not necessary

File: src/features/device-state/routes/state.ts [hunk 11]
```diff
@@ -95,7 +251,7 @@ const stateQuery = _.once(() =>
 					},
 				},
 				belongs_to__application: {
-					$select: ['id', 'app_name'],
+					$select: ['id', 'app_name', 'uuid'],
```
Comment (Page-) [inline src/features/device-state/routes/state.ts:254]: Functional change?

File: src/features/device-state/routes/state.ts (hunk 10 above)
Comment (toochevere) [reply src/features/device-state/routes/state.ts:72]: This is not a new change.  That code was there like that from before.   It only popped up now in the refactor because it was moved around.

File: src/features/device-state/routes/state.ts (hunk 9 above)
Comment (toochevere) [reply src/features/device-state/routes/state.ts:106]: removed

File: src/features/device-state/routes/state.ts [hunk 12]
```diff
@@ -157,241 +313,188 @@ const stateQuery = _.once(() =>
...
+
+	// grab the main app for this device...
+	return release == null
+		? {
+				name: userAppFromApi.app_name,
+				services: {},
+				networks: {},
+				volumes: {},
+				uuid: userAppFromApi.uuid,
+				appId: userAppFromApi.id,
```
Comment (Page-) [inline src/features/device-state/routes/state.ts:399]: These appear to be new to this PR?

File: src/features/device-state/routes/state.ts [hunk 13]
```diff
@@ -25,6 +26,161 @@ export const varListInsert = (varList: EnvVarList, obj: Dictionary<string>) => {
...
+
+	return {
+		releaseId: release.id,
+		commit: release.commit,
+		name: application.app_name,
+		services,
+		networks: composition?.networks || {},
+		volumes: composition?.volumes || {},
+		appId: application.id,
+		uuid: application.uuid,
```
Comment (Page-) [inline src/features/device-state/routes/state.ts:180]: These appear to be new to this PR?

File: src/features/device-state/routes/state.ts (hunk 10 above)
Comment (dfunckt) [reply src/features/device-state/routes/state.ts:72]: > although I don't think it's still non-nullable

@Page- it is still non-nullable, FYI -- I feel that `{}` is as good to represent "no composition" and less error-prone (ie. code always reads valid JSON).

File: src/features/device-state/routes/state.ts (hunk 13 above)
Comment (toochevere) [reply src/features/device-state/routes/state.ts:158]: The entire `type App` and method `buildAppFromRelease` are new to this PR as part of the refactoring.  `appId` is being used in this refactored code, `uuid` is not yet being used.  I can remove `uuid`.

File: src/features/device-state/routes/state.ts (hunk 12 above)
Comment (toochevere) [reply src/features/device-state/routes/state.ts:377]: see other comment

File: src/features/device-state/routes/state.ts [hunk 14]
```diff
@@ -25,6 +26,137 @@ export const varListInsert = (varList: EnvVarList, obj: Dictionary<string>) => {
...
+	});
+
+	return {
+		releaseId: release.id,
+		commit: release.commit,
+		name: application.app_name,
+		services,
+		networks: composition?.networks || {},
+		volumes: composition?.volumes || {},
+		appId: application.id,
```
Comment (Page-) [inline src/features/device-state/routes/state.ts:156]: This looks to be new?

File: src/features/device-state/routes/state.ts [hunk 15]
```diff
@@ -157,241 +289,187 @@ const stateQuery = _.once(() =>
...
+	const release = getReleaseForDevice(device);
+
+	// grab the main app for this device...
+	return release == null
+		? {
+				name: userAppFromApi.app_name,
+				services: {},
+				networks: {},
+				volumes: {},
+				appId: userAppFromApi.id,
```
Comment (Page-) [inline src/features/device-state/routes/state.ts:374]: This looks to be new?

File: src/features/device-state/routes/state.ts (hunk 14 above)
Comment (toochevere) [reply src/features/device-state/routes/state.ts:156]: No, it isn't.  It's just part of the refactoring.

File: src/features/device-state/routes/state.ts (hunk 15 above)
Comment (toochevere) [reply src/features/device-state/routes/state.ts:374]: Removing.

Comment (Page-) [reply src/features/device-state/routes/state.ts:373]: It's still here btw

Comment (toochevere) [reply src/features/device-state/routes/state.ts]: this time for real

**Review by Page- — APPROVED** [review]

**Comment by Page-** [comment]
@balena-ci rebase
